File: apps/website/frontend/src/types/models/registry.ts
// Virtual Arsenal registry catalog (T-068). Types only — the palette wiring lands in T-068.3.

/**
 * The kind of a registry item (T-068.10.2 v3 vocabulary; Phase 1 kinds remain
 * valid and gear_uniform is retired to 0 rows but still accepted). Mirrors
 * registry-items.schema.json#/$defs/item kind enum; taxonomy map:
 * .ai/artifacts/ace_arsenal_taxonomy_map.md.
 *
 * @model models.RegistryItem
 */
export type RegistryItemKind =
  | 'character'
  | 'gear_primary'
  | 'gear_handgun'
  | 'gear_launcher'
  | 'gear_throwable'
  | 'gear_explosive'
  | 'gear_uniform'
  | 'gear_jacket'
  | 'gear_pants'
  | 'gear_boots'
  | 'gear_vest'
  | 'gear_armored_vest'
  | 'gear_helmet'
  | 'gear_backpack'
  | 'gear_glasses'
  | 'gear_gloves'
  | 'gear_binoculars'
  | 'gear_item'
  | 'magazine'
  | 'ammo'
  | 'optic'
  | 'attachment'
  | 'vehicle'
  | 'vehicle_weapon'
  | 'crate'
  | 'other'

/**
 * One Virtual Arsenal catalog item, identified by its full Enfusion resource_name.
 *
 * @model models.RegistryItem
 * @contract registry-items.schema.json#/$defs/item
 */
export interface RegistryItem {
  id: string
  modpack_id: string
  resource_name: string
  display_name: string
  category: string
  icon_url?: string | null
  kind: RegistryItemKind
  /** Non-placeable template prefab (`*_base.et` / `* Base`) — hidden from pickers. */
  abstract?: boolean | null
  /** SCR_EArsenalItemType flag name when a faction EntityCatalog entry exists. */
  arsenal_type?: string | null
  /** ItemPhysicalAttributes.Weight in kg; null = engine class default. */
  weight_kg?: number | null
  /** ItemPhysicalAttributes.ItemVolume in cm³; null = engine class default. */
  volume_cm3?: number | null
  /** Container carry capacity (kg) when the item is itself a container. */
  max_weight_kg?: number | null
  /** Container volume capacity (cm³) when the item is itself a container. */
  max_volume_cm3?: number | null
  /** Addon ID the prefab was scanned from (joins the envelope addons[] scan set). */
  addon?: string | null
  sort_order: number
  created_at: string
  updated_at: string
}

/**
 * The registry list response: catalog rows plus the modpack id/version and a weak ETag.
 *
 * @contract registry-items.schema.json#/
 * @route GET /api/v1/registry
 */
export interface RegistryResponse {
  data: RegistryItem[]
  etag: string
  modpack_id: string
  modpack_version: string
}

/**
 * A compat edge family (T-150 v1 vocabulary). The graph code treats edge types
 * as plain strings so future families flow through untouched; this union is
 * the current schema enum for call-site safety.
 *
 * @contract registry-compat.schema.json#/$defs/edge/properties/edge_type
 */
export type RegistryCompatEdgeType =
  | 'mag_in_weapon'
  | 'ammo_in_mag'
  | 'optic_on_weapon'
  | 'attachment_on_weapon'
  | 'mag_in_vehicle_weapon'
  | 'ammo_in_vehicle_weapon'
  | 'character_default_loadout'
  | 'character_default_weapon'

/**
 * One directed compatibility edge: `from_node` goes in/on `to_node`.
 * `evidence` is omitted by the API when empty (NULL ≡ '' ≡ absent).
 *
 * @model models.RegistryCompatEdge
 * @contract registry-compat.schema.json#/$defs/edge
 */
export interface RegistryCompatEdge {
  id: string
  modpack_id: string
  from_node: string
  to_node: string
  edge_type: RegistryCompatEdgeType
  evidence?: string | null
  created_at: string
  updated_at: string
}

/**
 * The compat graph response: edge rows plus the modpack id/version and a weak
 * ETag (`?edge_type=` filtered responses carry a distinct ETag).
 *
 * @contract registry-compat.schema.json#/
 * @route GET /api/v1/registry/compat
 */
export interface RegistryCompatResponse {
  data: RegistryCompatEdge[]
  etag: string
  modpack_id: string
  modpack_version: string
}

File: apps/website/src/models/registry.rs
//! Registry model — Rust port of `internal/models/registry.go`.
//!
//! @contract registry-items.schema.json#/$defs/item

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::models::serde_helpers::go_time;

/// One placeable/equipable engine item in a modpack's flat catalog. Unique per
/// `(modpack_id, resource_name)`; `kind` holds the registry-items schema kind
/// vocabulary (v3, T-068.10.2) as plain text — new kinds need no model/DDL change.
/// v3 metadata columns are nullable: v2 envelopes leave them NULL.
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct RegistryItem {
    pub id: Uuid,
    pub modpack_id: Uuid,
    pub resource_name: String,
    pub display_name: String,
    pub category: String,
    #[serde(skip_serializing_if = "String::is_empty", default)]
    pub icon_url: String,
    pub kind: String,
    /// Non-placeable template prefab (`*_base.et` / `* Base`) — hidden from pickers.
    #[serde(rename = "abstract", skip_serializing_if = "Option::is_none", default)]
    #[sqlx(rename = "abstract")]
    pub abstract_: Option<bool>,
    /// SCR_EArsenalItemType flag name when the item has a faction EntityCatalog entry.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub arsenal_type: Option<String>,
    /// ItemPhysicalAttributes.Weight (kg); NULL = engine class default (not serialized).
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub weight_kg: Option<f64>,
    /// ItemPhysicalAttributes.ItemVolume (cm³); NULL = engine class default.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub volume_cm3: Option<f64>,
    /// Container carry capacity (m_fMaxWeight, kg) when the item is itself a container.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub max_weight_kg: Option<f64>,
    /// Container volume capacity (MaxCumulativeVolume, cm³) when the item is a container.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub max_volume_cm3: Option<f64>,
    /// Addon ID this prefab was scanned from (joins the envelope addons[] scan set).
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub addon: Option<String>,
    pub sort_order: i64,
    #[serde(with = "go_time")]
    pub created_at: DateTime<Utc>,
    /// Feeds the weak ETag (max updated_at).
    #[serde(with = "go_time")]
    pub updated_at: DateTime<Utc>,
}

/// One directed compatibility edge: `from_node` is the item that goes in/on,
/// `to_node` the host that accepts it. Unique per `(modpack_id, from_node,
/// to_node, edge_type)`; `edge_type` holds the registry-compat schema edge
/// vocabulary as plain text — new edge families need no model/DDL change.
///
/// @contract registry-compat.schema.json#/$defs/edge
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct RegistryCompatEdge {
    pub id: Uuid,
    pub modpack_id: Uuid,
    pub from_node: String,
    pub to_node: String,
    pub edge_type: String,
    #[serde(skip_serializing_if = "String::is_empty", default)]
    pub evidence: String,
    #[serde(with = "go_time")]
    pub created_at: DateTime<Utc>,
    /// Feeds the weak ETag (max updated_at).
    #[serde(with = "go_time")]
    pub updated_at: DateTime<Utc>,
}
